packages/go_router/lib/src/delegate.dart

```diff
@@ -150,20 +150,43 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
   bool _handlePopPageWithRouteMatch(
       Route<Object?> route, Object? result, RouteMatch? match) {
-    if (!route.didPop(result)) {
-      return false;
+    if (route.willHandlePopInternally) {
+      final bool popped = route.didPop(result);
+      assert(!popped);
+      return popped;
     }
     assert(match != null);
+    if (match!.route.onExit != null) {
+      // The `onExit` may perform updates to the navigator, such as popping up a
+      // dialog. Use a microtask to ensure the navigator finishes the current
+      // update before calling the `onExit`.
+      scheduleMicrotask(() async {
+        final bool onExitResult =
+            await match.route.onExit!(navigatorKey.currentContext!);
+        if (onExitResult) {
+          _removeMatchFromList(match, result);
+        }
+      });
+      return false;
+    } else {
+      if (!route.didPop(result)) {
+        return false;
+      }
+      _removeMatchFromList(match, result);
+      return true;
+    }
+  }
+
+  void _removeMatchFromList(RouteMatch match, Object? result) {
     if (match is ImperativeRouteMatch) {
```

Comment:
This method should check whether the match.completer is already completed

Reply:
I added a check in [Do not complete match.completer twice](https://github.com/flutter/packages/pull/3709/commits/01a0e0cdd8de74e61af36b17760b85d11eb6ebda)